Dockerfile

```
@@ -24,7 +24,28 @@ ENV VITE_GIT_COMMIT=${GIT_COMMIT}
 ENV VITE_BUILD_TIME=${BUILD_TIME}
 RUN npm run build
 
-# Stage 2: Python backend with frontend static files
+# Stage 2: Build the Python virtualenv with uv
+FROM python:3.12-slim AS python-builder
+
+WORKDIR /app
+
+# Install uv for faster dependency management
+RUN pip install uv
```

Hadolint (2.14.0)

[warning] Pin versions in pip. Instead of `pip install <package>` use `pip install <package>==<version>` or `pip install --requirement <requirements file>`

(DL3013)